File: src/hoodtrade/checks/contract.py
"""Contract-level safety checks.

Robinhood Chain launched with FCFS sequencing (which structurally blunts gas-auction
sandwiching), but it is permissionless and brand new — anyone can deploy, and the
token you are about to buy may be days old. These checks look at the token contract
itself, independent of the pool.
"""

from __future__ import annotations

from ..models import CheckResult, Severity
from ..rpc import decode_address, decode_string, decode_uint
from .base import Context

# EOA / empty code = not a contract at all. Buying a "token" with no code is a
# guaranteed loss (there is nothing to sell back).
ZERO = "0x" + "0" * 40


class ContractExistsCheck:
    id = "CONTRACT-EXISTS"

    async def run(self, ctx: Context) -> list[CheckResult]:
        code = await ctx.rpc.get_code(ctx.request.token)
        ctx.cache["token_code_size"] = len(code)
        if len(code) == 0:
            return [
                CheckResult(
                    check=self.id,
                    severity=Severity.DANGER,
                    score=100,
                    title="Token address has no contract code",
                    detail=(
                        "The token address is an externally-owned account or an undeployed "
                        "address. There is no ERC-20 to trade; a swap here cannot be reversed."
                    ),
                    evidence={"code_size": "0"},
                )
            ]
        return [
            CheckResult(
                check=self.id,
                severity=Severity.OK,
                score=0,
                title="Contract code present",
                detail=f"Token has {len(code)} bytes of deployed code.",
                evidence={"code_size": str(len(code))},
            )
        ]


class OwnershipCheck:
    id = "CONTRACT-OWNER"

    async def run(self, ctx: Context) -> list[CheckResult]:
        if ctx.cache.get("token_code_size") == 0:
            return []  # nothing to inspect
        owner = ZERO
        for fn in ("owner", "getOwner"):
            try:
                from ..rpc import encode_call

                raw = await ctx.rpc.eth_call(ctx.request.token, encode_call(fn))
                addr = decode_address(raw)
                if addr != ZERO:
                    owner = addr
                    break
            except Exception:  # noqa: BLE001 — absence of owner() is a finding, not an error
                continue

        if owner == ZERO:
            return [
                CheckResult(
                    check=self.id,
                    severity=Severity.OK,
                    score=0,
                    title="No active owner / renounced",
                    detail="No owner() found or ownership is renounced. Admin-key rug risk is lower.",
                )
            ]
        return [
            CheckResult(
                check=self.id,
                severity=Severity.WARN,
                score=15,
                title="Token has an active owner",
                detail=(
                    "An owner address can often pause transfers, mint supply, or change fees. "
                    "Not inherently malicious, but it is a live admin key you are trusting."
                ),
                evidence={"owner": owner},
            )
        ]


class SupplySanityCheck:
    id = "CONTRACT-SUPPLY"

    async def run(self, ctx: Context) -> list[CheckResult]:
        if ctx.cache.get("token_code_size") == 0:
            return []
        from ..rpc import encode_call

        try:
            name = decode_string(await ctx.rpc.eth_call(ctx.request.token, encode_call("name")))
            symbol = decode_string(await ctx.rpc.eth_call(ctx.request.token, encode_call("symbol")))
            supply = decode_uint(await ctx.rpc.eth_call(ctx.request.token, encode_call("totalSupply")))
        except Exception as exc:  # noqa: BLE001
            return [
                CheckResult(
                    check=self.id,
                    severity=Severity.WARN,
                    score=20,
                    title="Token does not implement standard ERC-20 reads",
                    detail=(
                        "name()/symbol()/totalSupply() did not decode. A non-standard token "
                        "can behave unpredictably in AMMs. Reason: " + str(exc)
                    ),
                )
            ]
        ctx.cache["token_symbol"] = symbol
        ctx.cache["token_name"] = name
        if supply == 0:
            return [
                CheckResult(
                    check=self.id,
                    severity=Severity.WARN,
                    score=15,
                    title="Zero total supply",
                    detail="totalSupply() is 0. Unusual for a live, tradable token.",
                    evidence={"name": name, "symbol": symbol},
                )
            ]
        return [
            CheckResult(
                check=self.id,
                severity=Severity.OK,
                score=0,
                title=f"Standard ERC-20: {symbol}",
                detail=f"{name} ({symbol}), total supply {supply}.",
                evidence={"name": name, "symbol": symbol, "total_supply": str(supply)},
            )
        ]

File: packages/scanner/src/checks/contract.ts
/**
 * Contract-level checks: does the token exist as code, is ownership renounced,
 * is the supply sane. Ported from Hood Trade (MIT).
 */

import type { CheckResult, TokenSnapshot } from "../types.js";

export function chainIdentity(s: TokenSnapshot): CheckResult[] {
  if (s.chainId == null || s.expectedChainId == null) return [];
  if (s.chainId !== s.expectedChainId) {
    return [
      {
        check: "CHAIN-IDENTITY",
        severity: "danger",
        score: 40,
        title: "RPC reports an unexpected chain id",
        detail: `The endpoint returned chain id ${s.chainId} but ${s.expectedChainId} was expected. You may be scanning the wrong network.`,
        evidence: { got: String(s.chainId), expected: String(s.expectedChainId) },
      },
    ];
  }
  return [{ check: "CHAIN-IDENTITY", severity: "ok", score: 0, title: "Chain identity confirmed", detail: `Connected to chain id ${s.chainId}.` }];
}

export function contractExists(s: TokenSnapshot): CheckResult[] {
  if (s.codeSize == null) return [];
  if (s.codeSize === 0) {
    return [
      {
        check: "CONTRACT-EXISTS",
        severity: "danger",
        score: 80,
        title: "No contract code at this address",
        detail: "The token address holds no bytecode. It is an EOA or an empty address — there is nothing to trade.",
      },
    ];
  }
  return [{ check: "CONTRACT-EXISTS", severity: "ok", score: 0, title: "Contract code present", detail: `${s.codeSize} bytes of bytecode at the token address.` }];
}

export function ownership(s: TokenSnapshot): CheckResult[] {
  if (s.ownerRenounced == null) return [];
  if (!s.ownerRenounced) {
    return [
      {
        check: "CONTRACT-OWNER",
        severity: "warn",
        score: 12,
        title: "Ownership not renounced",
        detail: "The contract still has an active owner. Depending on the code, the owner may be able to mint, pause, or change fees. Renounced ownership removes that discretion.",
      },
    ];
  }
  return [{ check: "CONTRACT-OWNER", severity: "ok", score: 0, title: "Ownership renounced", detail: "No privileged owner can alter the contract." }];
}

export function supplySanity(s: TokenSnapshot): CheckResult[] {
  if (s.totalSupply == null) return [];
  if (s.totalSupply <= 0) {
    return [
      {
        check: "CONTRACT-SUPPLY",
        severity: "danger",
        score: 30,
        title: "Total supply is zero",
        detail: "totalSupply() returned 0 — the token has no issued supply and cannot be traded normally.",
      },
    ];
  }
  return [{ check: "CONTRACT-SUPPLY", severity: "ok", score: 0, title: "Supply looks sane", detail: `totalSupply is ${s.totalSupply.toLocaleString()}.` }];
}
